src/search/pdbs/pattern_collection_generator_multiple.cc

```diff
@@ -0,0 +1,310 @@
+      rng(utils::parse_rng_from_options(opts)),
+      random_seed(opts.get<int>("random_seed")),
+      remaining_collection_size(opts.get<int>("max_collection_size")),
+      blacklisting(false),
+      stagnation_start_time(-1) {
+}
+
+void PatternCollectionGeneratorMultiple::check_blacklist_trigger_timer(
+    double blacklisting_start_time, const utils::CountdownTimer &timer) {
+    // Check if blacklisting should be started.
+    if (!blacklisting && timer.get_elapsed_time() > blacklisting_start_time) {
+        blacklisting = true;
+        // Also reset stagnation timer in case it was already set.
+        stagnation_start_time = -1;
+        if (verbosity >= utils::Verbosity::NORMAL) {
+            utils::g_log << "given percentage of total time limit "
+                         << "exhausted; enabling blacklisting." << endl;
+        }
+    }
+}
+
+unordered_set<int> PatternCollectionGeneratorMultiple::get_blacklisted_variables(
+    vector<int> &non_goal_variables) {
+    unordered_set<int> blacklisted_variables;
+    if (blacklisting && !non_goal_variables.empty()) {
+        /*
+          Randomize the number of non-goal variables for blacklisting.
+          We want to choose at least 1 non-goal variable and up to the
+          entire set of non-goal variables.
+        */
+        int blacklist_size = (*rng)(non_goal_variables.size());
+        ++blacklist_size; // [1, |non-goal variables|]
+        rng->shuffle(non_goal_variables);
+        blacklisted_variables.insert(
+            non_goal_variables.begin(), non_goal_variables.begin() + blacklist_size);
+        if (verbosity >= utils::Verbosity::DEBUG) {
+            utils::g_log << "blacklisting " << blacklist_size << " out of "
+                         << non_goal_variables.size()
+                         << " non-goal variables: ";
+            for (int var : blacklisted_variables) {
+                utils::g_log << var << ", ";
+            }
+            utils::g_log << endl;
+        }
+    }
+    return blacklisted_variables;
+}
+
+void PatternCollectionGeneratorMultiple::handle_generated_pattern(
+    PatternInformation &&pattern_info,
+    set<Pattern> &generated_patterns,
+    shared_ptr<PDBCollection> &generated_pdbs,
+    const utils::CountdownTimer &timer) {
+    const Pattern &pattern = pattern_info.get_pattern();
+    if (verbosity >= utils::Verbosity::DEBUG) {
+        utils::g_log << "generated pattern " << pattern << endl;
+    }
+    if (generated_patterns.insert(move(pattern)).second) {
+        // RCG generated a new pattern. Reset stagnation_start_time.
+        stagnation_start_time = -1;
+        shared_ptr<PatternDatabase> pdb = pattern_info.get_pdb();
+        remaining_collection_size -= pdb->get_size();
+        generated_pdbs->push_back(move(pdb));
+    } else {
+        // Pattern is not new. Set stagnation start time if not already set.
+        if (stagnation_start_time == -1) {
+            stagnation_start_time = timer.get_elapsed_time();
```

Comment:
I don't think it matters too much but it feels like the stagnation time is shifted by one evaluation: If I find pattern P 10 seconds into the algorithm, then find P again at 20 seconds, and then again at 30 seconds, I'd say, the method was stuck for 30 - 10 = 20 seconds. But here the time would start counting at 20, so it would consider the method not progressing for 20 - 30 = 10 seconds.

Reply:
I agree. Changing this, however, would mean to run experiments with the multiple CEGAR algorithm, too, so I'm not too eager to do this, but I could be persuaded.

Reply:
If it is possible to implement this without complicating the code too much, I'd be in favor of changing it but if it requires complex changes to keep track of the start value, I don't think it is necessary. If you leave it like this, I think this should be described in the documentation of the parameter because it doesn't do what you'd expect.
On the plus side, changing it could actually be beneficial, as a stagnation is recognized earlier.

Reply:
I implemented the change. If anything, I think that the logic got slightly easier. Please have a look at the commit.